File: frontend/src/api/backtest.ts
/**
 * Backtest admin route hooks + types (PR P2.4 / task #204).
 *
 * Mirrors `etfpulse/api/schemas/backtest.py` field-for-field. When the
 * backend schema changes, this file changes in the same PR — the type
 * shapes are the only contract between FE form/results panel and the
 * orchestrator output.
 *
 * Why a feature-scoped module instead of appending to `api/types.ts`
 * and `api/queries.ts`: both files are already past the 200–300 LOC
 * cap (599 and 653 respectively). Future feature work should follow
 * the same per-feature seam.
 */

import { useMutation, useQuery } from '@tanstack/react-query';

import { apiGet, apiPost } from './client';

/* ─────────────────────────── types ─────────────────────────── */

/** Allowed scalar types for detector kwargs. Mirrors the Python
 *  `DetectorKwargValue = bool | int | float | str` union. TypeScript
 *  doesn't distinguish int/float — `number` covers both. */
export type DetectorKwargValue = boolean | number | string;

/** Per-bucket scoring rubric tag. Mirrors `ScoringVersion` in the
 *  backend schema. */
export type BacktestScoringVersion = 'v2' | 'market-v1';

/** Request body for `POST /api/admin/backtest`. */
export interface BacktestRequest {
  /** ISO date YYYY-MM-DD. */
  start: string;
  /** ISO date YYYY-MM-DD. */
  end: string;
  /** Per-detector kwarg overrides. Unset = use production defaults. */
  detector_overrides?: Record<string, Record<string, DetectorKwargValue>>;
  /** Opt-in to live AI on cache miss. Requires `X-Backtest-Allow-AI:
   *  yes` header in addition to the body flag. */
  allow_ai?: boolean;
}

/** One outcome row in the report. Mirrors
 *  `pipeline.backtest.BacktestOutcomeRow` field-for-field. */
export interface BacktestOutcomeRow {
  detector_name: string;
  signal_type: string;
  asset: string;
  signal_date: string;
  fingerprint: string;
  direction: string | null;
  confidence: number | null;
  hit_target: boolean | null;
  hit_stop: boolean | null;
  composite_return_pct: string | null;
  scoring_version: BacktestScoringVersion | null;
  window_hours: number | null;
  skip_reason: string | null;
}

/** One per-detector summary row. Mirrors
 *  `pipeline.backtest.BacktestPerDetector`. */
export interface BacktestPerDetector {
  detector_name: string;
  n_hits: number;
  n_scored: number;
  wins: number;
  losses: number;
  /** null when n_scored == 0 — keeps null vs 0.0 distinguishable so
   *  the FE renders "—" instead of "0.0%". */
  hit_rate: number | null;
}

/** Full report shape returned by `POST /api/admin/backtest`. */
export interface BacktestReport {
  start: string;
  end: string;
  ai_prompt_version: string;
  detector_configs: Record<string, Record<string, DetectorKwargValue>>;
  counters: Record<string, number>;
  per_detector: BacktestPerDetector[];
  outcomes: BacktestOutcomeRow[];
}

/** One constructor kwarg for a detector. The FE form uses `type_name`
 *  to pick an input widget; `default` is rendered as the initial
 *  value when present. */
export interface BacktestDetectorParam {
  name: string;
  /** "int" | "float" | "Decimal" | "bool" | "str" — FE picks input
   *  widget based on this. */
  type_name: string;
  has_default: boolean;
  default: DetectorKwargValue | null;
}

/** One detector entry from `GET /api/admin/backtest/detectors`. */
export interface BacktestDetector {
  name: string;
  signal_type: string;
  params: BacktestDetectorParam[];
}

/** Response shape of `GET /api/admin/backtest/detectors`. */
export interface BacktestDetectorsResponse {
  detectors: BacktestDetector[];
}

/* ─────────────────────────── hooks ─────────────────────────── */

/** Cache-key namespace for backtest queries — keeps invalidation
 *  scopes narrow without colliding with the existing admin/metrics
 *  scope. */
const KEY_BACKTEST_DETECTORS = ['backtest', 'detectors'] as const;

/**
 * List backtest-eligible detectors with their constructor kwarg
 * signatures. The registry only changes on deploy, so we cache
 * indefinitely (`staleTime: Infinity`) and hold in memory for the
 * tab's lifetime (`gcTime: Infinity`).
 *
 * The hook is disabled while `adminKey` is empty so we don't fire a
 * request with no auth — mirrors the convention `useAdminMetrics`
 * uses for the same reason.
 */
export function useBacktestDetectors(adminKey: string) {
  return useQuery({
    queryKey: [...KEY_BACKTEST_DETECTORS, adminKey],
    queryFn: () =>
      apiGet<BacktestDetectorsResponse>(
        '/api/admin/backtest/detectors',
        undefined,
        { 'X-Admin-Key': adminKey },
      ),
    enabled: adminKey.length > 0,
    staleTime: Infinity,
    gcTime: Infinity,
    retry: false,
  });
}

/**
 * Run a backtest sweep. Write-shaped UX (operator clicks "Run" →
 * waits → sees result) even though the server is read-only, so a
 * mutation hook matches the interaction better than a parameterised
 * query. Retry off — a 4xx/5xx is either bad input or a real
 * orchestrator failure; re-firing wastes a request worker for 10–60s.
 *
 * `allow_ai=true` requires both the body flag and the
 * `X-Backtest-Allow-AI: yes` header. The hook always sends the header
 * when `request.allow_ai` is true so the operator's body+confirm flow
 * doesn't need to surface the header step manually.
 */
export function useBacktest(adminKey: string) {
  return useMutation({
    mutationFn: (request: BacktestRequest) => {
      const headers: Record<string, string> = { 'X-Admin-Key': adminKey };
      if (request.allow_ai) {
        headers['X-Backtest-Allow-AI'] = 'yes';
      }
      return apiPost<BacktestReport>('/api/admin/backtest', request, headers);
    },
    retry: false,
  });
}

File: backend/etfpulse/api/schemas/backtest.py
"""Response shapes for `POST /api/admin/backtest` (PR P2.4).

Mirrors the dataclasses in `pipeline.backtest` so a JSON-serialized
`BacktestReport.to_json_dict()` validates cleanly against the response
schema and the FE has a typed contract. The dataclasses stay the source
of truth for what the pipeline produces; this module is the on-the-wire
translation only.

Why scalar-union instead of `Any` for `detector_configs` and
`detector_overrides`:
  Detector constructor kwargs are heterogeneous SCALARS (int, float,
  bool, Decimal-as-string) — never nested objects. Constraining the
  union to those primitives keeps Pydantic validation tight and avoids
  the project's "no `Any` unless absolutely necessary" discipline.
  `Decimal` values arrive as strings (see `_build_detector` in
  `pipeline/backtest.py`) so `str` covers them losslessly.

Why `start`/`end` are strings (not `date`):
  The pipeline dataclass already serializes them as ISO strings (line
  144 of `pipeline/backtest.py`). Mirroring the wire shape keeps the
  schema lossless against `BacktestReport.to_json_dict()`. The request
  schema (`BacktestRequest`) uses `date` because Pydantic auto-parses
  ISO input there; route logic does the date→str conversion before
  calling the orchestrator.
"""

from __future__ import annotations

from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# Allowed kwarg value types for detector overrides. Mirrors the
# heterogeneous-scalar reality of detector constructor signatures.
# Pydantic's union resolution will coerce in-order; bool BEFORE int so
# `True` doesn't deserialize as `1`.
DetectorKwargValue = bool | int | float | str

# Literal mirroring `pipeline.backtest.BacktestOutcomeRow.scoring_version`.
ScoringVersion = Literal["v2", "market-v1"]


class BacktestRequest(BaseModel):
    """Body shape for `POST /api/admin/backtest`.

    Window-size cap and detector-override validation happen at the
    ROUTE layer (#201). This schema just pins shape + parses dates.
    """

    model_config = ConfigDict(extra="forbid")

    start: date
    end: date
    detector_overrides: dict[str, dict[str, DetectorKwargValue]] | None = None
    allow_ai: bool = False


class BacktestOutcomeRowOut(BaseModel):
    """One row in `BacktestReport.outcomes`. Mirrors
    `pipeline.backtest.BacktestOutcomeRow` field-for-field."""

    model_config = ConfigDict(extra="forbid")

    detector_name: str
    signal_type: str
    asset: str
    signal_date: str  # ISO date — pipeline serializes as str
    fingerprint: str
    direction: str | None
    confidence: int | None = Field(default=None, ge=1, le=10)
    hit_target: bool | None
    hit_stop: bool | None
    composite_return_pct: str | None  # str(Decimal) for lossless JSON
    scoring_version: ScoringVersion | None
    window_hours: int | None = Field(default=None, ge=0)
    skip_reason: str | None


class BacktestPerDetectorOut(BaseModel):
    """One row in `BacktestReport.per_detector`. Mirrors
    `pipeline.backtest.BacktestPerDetector`."""

    model_config = ConfigDict(extra="forbid")

    detector_name: str
    n_hits: int = Field(ge=0)
    n_scored: int = Field(ge=0)
    wins: int = Field(ge=0)
    losses: int = Field(ge=0)
    hit_rate: float | None = Field(default=None, ge=0.0, le=1.0)


class BacktestReportOut(BaseModel):
    """Response shape for `POST /api/admin/backtest`. Mirrors
    `pipeline.backtest.BacktestReport.to_json_dict()` exactly."""

    model_config = ConfigDict(extra="forbid")

    start: str  # ISO date
    end: str  # ISO date
    ai_prompt_version: str
    detector_configs: dict[str, dict[str, DetectorKwargValue]]
    counters: dict[str, int]
    per_detector: list[BacktestPerDetectorOut]
    outcomes: list[BacktestOutcomeRowOut]


class BacktestDetectorParamOut(BaseModel):
    """Single constructor kwarg signature for the detector listing route
    (#202). FE form renders one input per parameter."""

    model_config = ConfigDict(extra="forbid")

    name: str
    # Annotation type as a human-readable string ("int", "float", "Decimal",
    # "bool"). FE picks input widget based on this. Decimal renders as a
    # numeric text input (no native decimal type in browsers).
    type_name: str
    has_default: bool
    default: DetectorKwargValue | None = None


class BacktestDetectorOut(BaseModel):
    """One detector entry for `GET /api/admin/backtest/detectors`."""

    model_config = ConfigDict(extra="forbid")

    name: str
    signal_type: str
    params: list[BacktestDetectorParamOut]


class BacktestDetectorsResponse(BaseModel):
    """Response shape for `GET /api/admin/backtest/detectors`."""

    model_config = ConfigDict(extra="forbid")

    detectors: list[BacktestDetectorOut]
